include/openthread/dns_client.h

```diff
@@ -0,0 +1,470 @@
+ * @param[in]  aServerSockAddr  A pointer to the server socket address.
+ * @param[in]  aServiceName     The service name to query for (MUST NOT be NULL).
+ * @param[in]  aCallback        A function pointer that shall be called on response reception or time-out.
+ * @param[in]  aContext         A pointer to arbitrary context information.
+ *
+ * @retval OT_ERROR_NONE        Query sent successfully. @p aCallback will be invoked to report the status.
+ * @retval OT_ERROR_NO_BUFS     Insufficient buffer to prepare and send query.
+ *
+ */
+otError otDnsClientBrowse(otInstance *        aInstance,
+                          const otSockAddr *  aServerSockAddr,
+                          const char *        aServiceName,
+                          otDnsBrowseCallback aCallback,
+                          void *              aContext);
+
+/**
+ * This function gets the service name associated with a DNS browse (service instance enumeration) response.
+ *
+ * This function MUST only be used from `otDnsBrowseCallback`.
+ *
+ * @param[in]  aResponse         A pointer to the response.
+ * @param[out] aNameBuffer       A buffer to char array to output the service name (MUST NOT be NULL).
+ * @param[in]  aNameBufferSize   The size of @p aNameBuffer.
+ *
+ * @retval OT_ERROR_NONE     The service name was read successfully.
+ * @retval OT_ERROR_NO_BUFS  The name does not fit in @p aNameBuffer.
+ *
+ */
+otError otDnsBrowseResponseGetServiceName(const otDnsBrowseResponse *aResponse,
+                                          char *                     aNameBuffer,
+                                          uint16_t                   aNameBufferSize);
+
+/**
+ * This function gets a service instance associated with a DNS browse (service instance enumeration) response.
+ *
+ * This function MUST only be used from `otDnsBrowseCallback`.
+ *
+ * The response may include multiple service instance records. @p aIndex can be used to iterate through the list. Index
+ * zero gives the the first record. When we reach end of the list, `OT_ERROR_NOT_FOUND` is returned.
+ *
+ * Note that this function gets the service instance label and not the full service instance name which is of the form
+ * `<Instance>.<Service>.<Domain>`.
+ *
+ * @param[in]  aResponse          A pointer to the response.
+ * @param[in]  aIndex             The service instance record index to retrieve.
+ * @param[out] aLabelBuffer       A buffer to char array to output the service instance label (MUST NOT be NULL).
+ * @param[in]  aLabelBufferSize   The size of @p aLabelBuffer.
+ *
+ * @retval OT_ERROR_NONE          The service instance was read successfully.
+ * @retval OT_ERROR_NO_BUFS       The name does not fit in @p aNameBuffer.
+ * @retval OT_ERROR_NOT_FOUND     No service instance record in @p aResponse at @p aIndex.
+ * @retval OT_ERROR_PARSE         Could not parse the records in the @p aResponse.
+ *
+ */
+otError otDnsBrowseResponseGetServiceInstance(const otDnsBrowseResponse *aResponse,
```

Comment:
I originally had it `GetInstanceLabel()` (I did not want to use a `Name` since we are getting the label only and not full name), later I changed it to `GetServiceInstance()`.

Still prefer not using `Name` here.

Thoughts?

Reply:
Makes sense to me to avoid confusing with full names. Thanks for sharing the thoughts!